projects/RabbitMQ.Client/client/api/CachedString.cs

```diff
@@ -0,0 +1,63 @@
+﻿using System;
+using System.Text;
+
+namespace RabbitMQ.Client
+{
+    /// <summary>
+    /// Caches a string's byte representation to be used for certain methods like <see cref="IModel.BasicPublish(CachedString,CachedString,bool,IBasicProperties,ReadOnlyMemory{byte})"/>.
+    /// </summary>
+    public sealed class CachedString
```

Comment:
@bollhals what is the benefit of  exposing this as method parameters vs exposing the CLR type ReadonlyMemory? At first glance it looks like only convenience right? So would it make sense to make the public Api just accepting the CLR type instead? I'm asking because even with cached string the caller of the API needs to be concerned with caching the instances so the value of this class is really only abstracting the encoding get bytes which I find not enough to expose a custom type. A middle ground could be to expose the type but make it convertible to readonly memory while still switching the public Api to readonly memory

Reply:
I hope my explanation is reasonable and what I wrote is understandable 😄 
This was one thing which I was not 100% certain what the best option might be. In the end I've exposed it as CachedString due to:
- You should only use this method if you actually cache the string (Otherwise you have worse performance)
- Exposing ROM enables customer to send non UTF8 values, which might cause more harm than good. (In which case would you have a valid ROM but not a string? (Stackalloc is also not possible since it's Memory not Span))
- In the future (comment from other PR) we could start returning the CachedString in other methods like BasicConsume or QueueDeclare, which then again could directly be used again in follow up methods for better performance.
- CachedString ist more clearly explaining the intent of the method than ROM<byte> + comments explaining it has to be a UTF8 sting

Let's switch it around, what impact would you get from exposing just ROM<byte>?
Which use cases could you see which taking a ROM<byte> is preferable than CachedString?

The intent of this API is better performance due to not having to encode the same string over and over again which is done by caching the encoded value, which implies that the user is caching the value.

Reply:
Ok got it. These are really the strongest reason to expose a custom type

> - You should only use this method if you actually cache the string (Otherwise you have worse performance)
> - Exposing ROM enables customer to send non UTF8 values, which might cause more harm than good. (In which case would you have a valid ROM but not a string? (Stackalloc is also not possible since it's Memory not Span))

> Which use cases could you see which taking a ROM is preferable than CachedString?

My thoughts were comming from the experience that when you have a .NET type available to expose a .NET type where it makes sense because it is one API less you have to maintain and potentially evolve. But like mentioned above the two major bulllets seem to indicate a strong custom type makes sense.

Regarding

> - CachedString ist more clearly explaining the intent of the method than ROM + comments explaining it has to be a UTF8 string

Could we add some validation into the constructor of that type or at least call it `CachedUTF8String` or something similar?

Reply:
> > CachedString ist more clearly explaining the intent of the method than ROM + comments explaining it has to be a UTF8 string

> Could we add some validation into the constructor of that type or at least call it CachedUTF8String or something similar?

What kind of validation?
Sure we can rename it, but I'm not favoring UTF8 in the name as the type kind of abstracting it away so you don't need to care what it is. Adding it in the name makes it clear what it will use, but for what purpose?